proposal/package-importer.md

```diff
@@ -0,0 +1,191 @@
+
+[Sass pkg: test]: https://github.com/oddbird/sass-pkg-test
+
+### Design Decisions
+
+We could use the `~` popularized by Webpack's `load-sass` format, but this has
+been deprecated since 2021. In addition, since this creates a URL that is
+syntactically a relative URL, it does not make it clear to the implementation or
+the reader where to find the file.
+
+While the Dart Sass implementation allows for the use of the `package:` url
+scheme, a similar standard doesn't exist in Node. We could add `node_modules` to
+the load path, but that again suffers from lack of clarity to the implementation
+and the reader.
+
+Some npm packages have adopted a convention of declaring their Sass entrypoints
+using `"style"` or `"sass"` keys in their `package.json`. This is tied to the
+npm context, and would require parsing package.json to find these files.
+Instead, each implementation should use the standard patterns and tooling
+(`require` for Node and `package_config` for Dart) for resolving file paths.
+This allows library authors control over what is public, which follows Sass's
+general design.
+
+This proposal does not add support for a [Node custom user condition]. Node
+currently does not provide a native way to set a condition on a single
+invocation of `require.resolve`, meaning we would need to implement our own
+`package.json` parsing similar to [Rollup] or add a dependency on a community
+solution like [resolve.exports]. However, this proposal will not prevent future
+support.
+
+[Node custom user condition]: https://nodejs.org/api/packages.html#community-conditions-definitions
+[Rollup]: https://github.com/rollup/plugins/blob/master/packages/node-resolve/src/package/resolvePackageExports.js
+[resolve.exports]: https://github.com/lukeed/resolve.exports
+
+A potential source of confusion is in instances where a dependency defines a
+default or main export that is not in the root folder. In these cases, Sass will
+look for index and partial files in the same folder as the default or main
+export, and not in the package root.
+
+The `pkg` scheme will not be supported in the browser version of Dart Sass. To
+support a similar functionality, a user would need to ensure that files are
+served, and the loader would need to fetch the URL. In order to follow the same
+algorithm for [resolving a file: URL], we would need to make many fetches. If we
+instead require the browser version to have a fully resolved URL, we negate many
+of this spec's benefits.
+
+[resolving a file: URL]: ../spec/modules.md#resolving-a-file-url
+
+## Semantics
+
+This proposal defines a new step in the [Loading a Source File] in modules that
+implementations must include.
+
+New step for [Loading a Source File], after "if argument is a relative URL"
+step:
+
+- If `argument` is a valid URL with scheme `pkg`:
+  - Let `resolved` be the result of resolving a pkg url
+  - If `resolved` is not null, let `argument` equal `resolved`, and continue.
+    Otherwise, return null.
```

Comment:
There are two ways we could go here for defining "package importers":

1. Reserve the URL scheme `pkg:` and dispatch it specially to a designated package importer, which is what this proposal does currently.
2. Allow _any importer_ to resolve `pkg:` imports and just provide an official recommendation for the structure of `pkg:` URLs as well as a built-in importer for Node.js semantics.

The benefits of option 1 are that you can give package importers some assistance in resolving URLs (as you do below), and that you can enforce more stringent requirements on them (like _only_ processing `pkg:` URLs) than you can on general importers.

On the other hand, option 2 gives authors a lot more flexibility. For example, you can imagine someone wanting to support `pkg:` URLs in the browser via [import maps](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap#import_map_json_representation), or even just from a virtual filesystem in web IDE. And authors who do want to target the filesystem can mitigate the pain by writing [`FileImporter`s](https://sass-lang.com/documentation/js-api/interfaces/fileimporter/) that handle the `pkg:` URL scheme.

So my recommendation is this: Don't define any new language semantics. Instead, define the _expectations_ for an importer that handles `pkg:` URLs: that the first component will be the package name, that it will automatically try to find an entrypoint file, that nested files can be addressed using `/`, and so on.